bpf/flows.c

```diff
@@ -95,59 +151,137 @@ static inline int fill_ip6hdr(struct ipv6hdr *ip, void *data_end, struct flow *f
+}
 
+static inline int flow_monitor (struct __sk_buff *skb, u8 direction) {
     // If sampling is defined, will only parse 1 out of "sampling" flows
     if (sampling != 0 && (bpf_get_prandom_u32() % sampling) != 0) {
         return TC_ACT_OK;
     }
-
-    void *data = (void *)(long)skb->data;
     void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    flow_id_v id;
+    int rc = TC_ACT_OK;
+    u32 flags = 0;
+    flow_record *flow_event;
+    flow_metrics *my_flow_counters;
+    __u64 current_time = bpf_ktime_get_ns();
 
-    struct flow *flow = bpf_ringbuf_reserve(&flows, sizeof(struct flow), 0);
-    if (!flow) {
+    struct ethhdr *eth = data;
+    if (fill_ethhdr(eth, data_end, &id, &flags) == DISCARD) {
         return TC_ACT_OK;
     }
+    if (direction == INGRESS) {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_ingress, &id);
+    } else {
+        my_flow_counters = bpf_map_lookup_elem(&xflow_metric_map_egress, &id);
+    }
 
-    struct ethhdr *eth = data;
-    if (fill_ethhdr(eth, data_end, flow) == DISCARD) {
-        bpf_ringbuf_discard(flow, 0);
+    if (my_flow_counters != NULL) {
+        my_flow_counters->packets += 1;
+        my_flow_counters->bytes += skb->len;
+        my_flow_counters->last_pkt_ts = current_time;
+        if (flags & TCP_FIN_FLAG || flags & TCP_RST_FLAG) {
+            /* Need to evict the entry and send it via ring buffer */
+            flow_event = bpf_ringbuf_reserve(&flows, sizeof(flow_record), 0);
+            if (!flow_event) {
+                return rc;
+            }
+            export_flow_id(&flow_event->flow_key, id, direction);
+            flow_event->metrics.flags = flags;
+            bpf_ringbuf_submit(flow_event, 0);
+            // Defer the deletion of the entry from the map to usespace since it evicts other CPU metrics
+        } else {
+            if (direction == INGRESS) {
+                bpf_map_update_elem(&xflow_metric_map_ingress, &id, my_flow_counters, BPF_EXIST);
+            } else {
+                bpf_map_update_elem(&xflow_metric_map_egress, &id, my_flow_counters, BPF_EXIST);
+            }
+        }
     } else {
-        flow->direction = direction;
-        flow->bytes = skb->len;
-        bpf_ringbuf_submit(flow, 0);
+        flow_metrics new_flow_counter = {
+            .packets = 1, .bytes=skb->len};
+        new_flow_counter.flow_start_ts = current_time;
+        new_flow_counter.last_pkt_ts = current_time;
+        int ret;
+        if (direction == INGRESS) {
+            ret = bpf_map_update_elem(&xflow_metric_map_ingress, &id, &new_flow_counter,
+                                      BPF_NOEXIST);
+        } else {
+            ret = bpf_map_update_elem(&xflow_metric_map_egress, &id, &new_flow_counter,
+                                      BPF_NOEXIST);
+        }
+        if (ret < 0) {
+            /*
+                When the map is full, we have two choices:
```

Comment:
There's probably an interesting discussion to have here.
I guess the two suggested options both involve a drop in performance. Moreover, when the map is full, we could assume there's high probability that it will stay full for a while (?) so it's like we're entering in a degraded mode.
What about other options:
1. Flushing the whole map when it's full, so that performance gains kinda "restart from scratch"
2. Just drop the flow?

In all cases, it would be nice to be able to report this kind of information to the ebpf agent (userspace) so that, maybe, it could emit some metrics about the state of the bpf maps? It could be useful to track performance gains and drops, to see if it correlates with maps being full ?

Reply:
Yes, currently, I constantly monitor the size of the maps, no. of collisions etc. Which in future, we can use to take wise decisions.